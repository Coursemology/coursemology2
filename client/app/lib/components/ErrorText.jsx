import PropTypes from 'prop-types';
import { red } from '@mui/material/colors';

/**
 * Standardises the way errors are shown in redux forms.
 */
const ErrorText = ({ errors }) => {
  if (!errors) {
    return null;
  }
  if (errors.constructor === String) {
    return <div style={{ color: red[500] }}>{errors}</div>;
  }
  if (errors.constructor === Array) {
    return (
      <>
        {errors.map((error) => (
          <ErrorText key={error} errors={error} />
        ))}
      </>
    );
  }
  return null;
};

export const errorProps = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.arrayOf(PropTypes.string),
]);

ErrorText.propTypes = {
  errors: errorProps,
};

export default ErrorText;
