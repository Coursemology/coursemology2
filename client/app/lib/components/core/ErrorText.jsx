import { FormattedMessage } from 'react-intl';
import { Typography } from '@mui/material';
import PropTypes from 'prop-types';

/**
 * Standardises the way errors are shown in redux/react-hook forms.
 */
const ErrorText = ({ errors }) => {
  if (
    !errors ||
    (errors.constructor === Object && Object.keys(errors).length === 0) ||
    (errors.constructor === Array && errors.length === 0)
  ) {
    return null;
  }
  if (errors.constructor === String) {
    return (
      <Typography color="error" variant="body2">
        {errors}
      </Typography>
    );
  }
  if (errors.constructor === Array) {
    return (
      <>
        {errors.map((error) => (
          <ErrorText key={error} errors={error} />
        ))}
      </>
    );
  }
  // When an error with the 'base' attribute is returned by RoR, show it.
  if (errors.constructor === Object && errors.base) {
    return <ErrorText key={errors.base} errors={errors.base.message} />;
  }
  return (
    <Typography color="error" variant="body2">
      <FormattedMessage
        defaultMessage="Failed submitting this form. Please try again."
        id="lib.components.core.ErrorText.error"
      />
    </Typography>
  );
};

export const errorProps = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.arrayOf(PropTypes.string),
  PropTypes.object,
]);

ErrorText.propTypes = {
  errors: errorProps,
};

export default ErrorText;
