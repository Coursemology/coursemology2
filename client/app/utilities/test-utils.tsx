// eslint-disable-next-line import/no-extraneous-dependencies
import { render, RenderResult } from '@testing-library/react';

import TestApp, { CustomRenderOptions } from './TestApp';

const customRender = (
  ui: JSX.Element,
  options?: CustomRenderOptions,
): RenderResult => render(<TestApp {...options}>{ui}</TestApp>);

// eslint-disable-next-line import/no-extraneous-dependencies
export * from '@testing-library/react';
export { customRender as render };
