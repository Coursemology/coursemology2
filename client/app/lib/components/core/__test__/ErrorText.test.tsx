import { render } from 'test-utils';

import ErrorText from '../ErrorText';

describe('<ErrorText />', () => {
  describe('when input is a string', () => {
    const errors = 'An error.';

    it('displays it', () => {
      expect(render(<ErrorText errors={errors} />)).toMatchSnapshot();
    });
  });

  describe('when input is an array', () => {
    const errors = ['An error.', 'Another error.'];

    it('displays each error', () => {
      expect(render(<ErrorText errors={errors} />)).toMatchSnapshot();
    });
  });
});
