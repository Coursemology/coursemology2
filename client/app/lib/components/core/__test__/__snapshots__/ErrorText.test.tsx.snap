// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorText /> when input is a string displays it 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="color: rgb(244, 67, 54);"
      >
        An error.
      </div>
      <div
        class="Toastify"
      />
    </div>
  </body>,
  "container": <div>
    <div
      style="color: rgb(244, 67, 54);"
    >
      An error.
    </div>
    <div
      class="Toastify"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<ErrorText /> when input is an array displays each error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="color: rgb(244, 67, 54);"
      >
        An error.
      </div>
      <div
        style="color: rgb(244, 67, 54);"
      >
        Another error.
      </div>
      <div
        class="Toastify"
      />
    </div>
  </body>,
  "container": <div>
    <div
      style="color: rgb(244, 67, 54);"
    >
      An error.
    </div>
    <div
      style="color: rgb(244, 67, 54);"
    >
      Another error.
    </div>
    <div
      class="Toastify"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
